경북대 BE 김진현 - Step3

kimjhyun0627 opened the pull request:
## 구현한 기능

- Step 2 코드리뷰 반영
   - `@PageableDefault`의 상수값들을 따로 객체로 분리한 이유는, 위시리스트의 경우 '내 위시리스트 보기' 뿐만 아니라 '나와 같은 상품을 담은 친구 보기' 등의 기능 확장 과정에서나, 아직 UI/UX가 제대로 정립되지 않은 상황에서 각 page에서 표시할 정보의 양과 정렬 순서를 손쉽게 변경하고자 함이었습니다. 멘토님 리뷰를 보고 응집성이 떨어짐을 확인했고, 관련 코드 리팩토링 진행했습니다.
   - `constant` 패키지의 경우 제 성격상 가까운 미래에 쓸 것 같아서...빈 패키지임에도 삭제하지 않았습니다..!
- 스키마 & 도메인 설계
   - 스키마는 `ON DELETE CASCADE`를 통해 상품 삭제 시 같이 삭제되도록 했습니다.
   - `Object Value`에서 이번에는 `Validation`을 사용한 예외처리를 최대한 활용해 보았습니다.
   - `static field`를 `public`으로 풀고 DTO 검증도 시도해보았지만...무슨 이유인지 `Exception Handling`이 잘 되지 않아 우선은 도메인 검증만 진행했습니다.
- 레포지토리 & 서비스 레이어 구현
   - 중복 검증을 위한 메소드 `existsByProduct_IdAndName_Name`을 추가했습니다.
- 컨트롤러 & DTO 구현
   - 사실 이 부분이 고민되었는데, 현재는 옵션을 추가하거나 조회할 상품의 인가 확인을 위해 `ProductOptionController`에서 기존 `ProductService`의 메소드를 사용하고 있습니다.
   - 이 방식이 역할 분리 원칙에 위배되지 않는지 여쭤보고 싶습니다.
   - PR을 진행하면서 생각한 대안은, `ProductOptionService`의 메소드가 파라미터 `role`을 받도록 리팩토링하고 `ProductOptionService`에서 `ProductService`를 불러와 처리하는 방식인데...
   - 마음은 후자로 기울긴 했지만, 사실 잘 모르겠습니다..! 조언 부탁드립니다 !
- 각각에 해당하는 단위 테스트 코드 작성
   - 기존 코드 참고해서 단위 테스트 코드를 작성했고, 이번에는 vo에 대한 단위 테스트도 만들어보았습니다.

kimjhyun0627 committed:
docs: step 2 코드리뷰 반영을 위한 README 업데이트

kimjhyun0627 committed:
refactor: Step 2 코드리뷰 반영: `@PageableDefault` 상수 인라인 처리

- `@PageableDefault` 상수 인라인 처리

kimjhyun0627 committed:
docs: README 업데이트

kimjhyun0627 committed:
feat: 도메인 및 스키마 설계

kimjhyun0627 committed:
feat: 레포지토리 & 서비스 레이어 구현

kimjhyun0627 committed:
feat: 컨트롤러 구현

kimjhyun0627 commented on src/main/java/gift/controller/user/ProductOptionController.java:
```diff
@@ -0,0 +1,77 @@
+        var role = extractRole(request);
+        productService.getProductById(productId, role)
+                .orElseThrow(() -> new ProductNotFoundException(productId));
+    }
```
PR 코맨트로 말씀드렸던 부분입니다 !

kimjhyun0627 replied:
PR 코맨트로 말씀드렸던 부분입니다 !

kimjhyun0627 commented on src/main/java/gift/controller/user/ProductOptionController.java:
```diff
@@ -0,0 +1,77 @@
+        return ResponseEntity.noContent().build();
+    }
+
+    private void authorizeAccess(Long productId, HttpServletRequest request) {
+        var role = extractRole(request);
+        productService.getProductById(productId, role)
+                .orElseThrow(() -> new ProductNotFoundException(productId));
+    }
```
PR 코멘트로 말씀드렸던 부분입니다 !

Gyuchool replied:
리뷰 반영해가면서 개선해가보시죠!

Gyuchool commented on src/main/java/gift/controller/admin/AdminProductController.java:
```diff
@@ -40,7 +37,7 @@ public AdminProductController(ProductService productService) {
     public String list(
             @CurrentRole Role role,
-            @PageableDefault(size = PRODUCT_PAGE_SIZE, sort = PRODUCT_PAGE_KEY, direction = Sort.Direction.ASC) Pageable pageable,
+            @PageableDefault(size = 15, sort = "id", direction = Sort.Direction.ASC) Pageable pageable,
```
리뷰 반영 감사합니다☺️

Gyuchool commented on src/main/java/gift/entity/product/option/ProductOption.java:
```diff
@@ -0,0 +1,84 @@
+    protected ProductOption() {
+    }
+
+    private ProductOption(OptionName name, OptionQuantity quantity) {
```
정적팩토리 메서드를 사용하기때문에 private 선언👍 좋은 습관이네요!!

Gyuchool commented on src/main/java/gift/entity/product/option/ProductOption.java:
```diff
@@ -0,0 +1,84 @@
+        this.product = product;
+    }
+
+    public void decrease(int amount) {
```
decrease만 네이밍을 넣으면 어떤걸 감소하는지 명확히 알 수 없는 것 같아요! 다른 수치 정보를 가진 필드가 추가되면 여기도 수정해야할 것 같습니다!
구체적인 네이밍 부탁드릴게요:)

kimjhyun0627 replied:
`decreaseAmount`로 변경했습니다!

Gyuchool commented on src/main/java/gift/entity/product/option/value/OptionName.java:
```diff
@@ -0,0 +1,22 @@
+import org.hibernate.validator.constraints.Length;
+
+@Embeddable
+public record OptionName(
```
validation을 이용한 검증과 생성자에서 검증하는 방법과 어떤 차이가 있을까요?🤔

kimjhyun0627 replied:
생성자 검증은 로직 안에서 검사를 하고, validation은 validator.validate() 호출을 통해 검증이 이루어진다고 학습했습니다 !
학습 과정에서 생성자 검증은 간단한 불변 조건에 해당해 생성자 검증이 적합하다고 생각해 이 부분 수정하겠습니다 !

Gyuchool commented on src/main/java/gift/service/product/option/ProductOptionServiceImpl.java:
```diff
@@ -0,0 +1,67 @@
+                new ProductId(productId),
+                name
+        );
+        if (exists) {
```
exists를 재사용하지 않는다면 바로 조건문에 감싸는게 가독성이 좋아보이네요!

Gyuchool commented on src/test/java/gift/entity/product/option/OptionNameTest.java:
```diff
@@ -0,0 +1,55 @@
+
+    private static Validator validator;
+
+    @BeforeAll
```
BeforeEach, BeforeAll의 차이가 무엇일까요?🙄

kimjhyun0627 replied:
`@BeforeEach`는 매 테스트마다, `@BeforeAll`은 테스트 클래스 실행 시 처음 한 번만 실행된다고 알고 있습니다 !

Gyuchool commented on src/test/java/gift/entity/product/option/OptionNameTest.java:
```diff
@@ -0,0 +1,55 @@
+    }
+
+    @Test
+    void validNameShouldHaveNoViolations() {
```
DisplayName이 있으면 테스트 의도를 이해하기 더 쉬울 것 같습니다!

Gyuchool commented on src/test/java/gift/service/product/ProductOptionServiceTest.java:
```diff
@@ -0,0 +1,178 @@
+@DisplayName("ProductOptionServiceImpl 단위 테스트")
+class ProductOptionServiceTest {
+
+    private final Long PRODUCT_ID = 1L;
+    private final ProductId PRODUCT_ID_WRAPPER = new ProductId(PRODUCT_ID);
```
상수는 static을 붙여서 선언하는데 static이 붙고 안붙고의 차이가 무엇일까요?🙄

kimjhyun0627 replied:
static을 붙여 모든 객체가 같은 주소값을 사용하도록 만듭니다 !

Gyuchool commented on src/test/java/gift/service/product/ProductOptionServiceTest.java:
```diff
@@ -0,0 +1,178 @@
+import org.mockito.Mock;
+import org.mockito.junit.jupiter.MockitoExtension;
+
+@ExtendWith(MockitoExtension.class)
```
Mock test👍
적용해보시면서 Mock 테스트와 통합 테스트의 차이점은 무엇이라고 생각하실까요?

kimjhyun0627 replied:
Mock 테스트는 DB와 같은 외부 의존성을 모두 Mock으로 대체해 서비스나 클래스의 개별 로직만 검증하는 반면, 통합 테스트는 실제 환경에서 애플리케이션을 띄워 서비스·리포지토리·DB 같은 여러 구성 요소가 함께 제대로 연동되는지를 확인합니다.

Gyuchool commented on src/test/java/gift/entity/product/option/OptionQuantityTest.java:
```diff
@@ -0,0 +1,61 @@
+
+    @Test
+    void aboveMaxShouldViolateMax() {
+        OptionQuantity qty = new OptionQuantity(100_000_000);
```
3개씩 잘라서 표현하는거 좋네요👍

Gyuchool commented on src/test/java/gift/controller/user/ProductOptionControllerTest.java:
```diff
@@ -0,0 +1,192 @@
+
+        given(optionService.getOptions(productId)).willReturn(options);
+
+        try (MockedStatic<RoleUtil> rs = Mockito.mockStatic(RoleUtil.class)) {
```
static 클래스로 인해서 테스트하기 어려워 보이네요🤔
위의 구조를 개선하면 여기도 자연스럽게 해소될 것으로 보입니다!

Gyuchool commented on src/main/java/gift/controller/user/ProductOptionController.java:
```diff
@@ -0,0 +1,77 @@
+import gift.exception.custom.ProductNotFoundException;
+import gift.service.product.ProductService;
+import gift.service.product.option.ProductOptionService;
+import jakarta.servlet.http.HttpServletRequest;
```
controller가 servlet을 알아야할까요? 🤔
spring을 사용하는 이유가 무엇인지 spring의 어떤 장점 때문에 사용하시는지 고민해보시죠!

kimjhyun0627 replied:
`AdminProductController`에서는 `resolve`를 사용하고 있었는데, REST API 부분은 놓치고 있었네요..! 수정하겠습니다 !

Gyuchool commented on src/main/java/gift/controller/user/ProductOptionController.java:
```diff
@@ -0,0 +1,77 @@
+    }
+
+    private void authorizeAccess(Long productId, HttpServletRequest request) {
+        var role = extractRole(request);
```
Util클래스로 빼서 사용하고 있었군요! util 클래스를 지양하는 것을 추천하는 이유가 여기서 나온는 것 같습니다!
Role을 꺼내는 책임을 적절한 곳에 넘기면 어떨까요? 그러면 공통적인 authorizeAccess()가 제거될 것 같네요!

kimjhyun0627 replied:
CurrentRole을 받는 resolver를 따로 만들고, 리졸버에서 처리하도록 변경했습니다! 사용되지 않는 파라미터는 _언더스코어 처리했습니다!

Gyuchool commented on src/main/java/gift/entity/product/option/value/OptionQuantity.java:
```diff
@@ -0,0 +1,28 @@
+        int quantity
+) {
+
+    private static final int MAX_OPT_QUANTITY_SIZE = 99_999_999;
```
99_999_999보단 1억이 읽기 쉬울 것 같은데 100_000_000으로 수정은 어떠실까요? 
MAX value값때문이라면 -1 더해서 해결할 수 있을 것 같습니다:)

Gyuchool reviewed (changes requested):
안녕하세요 진현님!
마지막 3단계 미션 진행하시느라 고생하셨습니다!! 👍 빠르게 구현해주셨네요👍

요구사항대로 잘 구현해주셔서 머지해도 괜찮겠지만 더 고민하고 학습할만한  포인트가 있어서 RC드립니다💯

몇 가지 코멘트를 남겨두었으니 반영 및 의견 주시고 다시 요청주세요!🙏
궁금한 점은 디스코드 DM이나 PR 커멘트로 언제든지 여쭤봐주세요!(새벽도 가능!)
끝까지 화이팅입니다💪

kimjhyun0627 committed:
refactor: Step 3 리뷰 반영

- `ProductOption.java`: `decrease` 메소드 네이밍 변경
- `entity/product/option`: 도메인 VO 검증 방식 번경
- `ProductOptionServiceImpl.java`: `exists` 인라인화
- `ProductController.java`: resolver 사용하도록 변경
- `ProductOptionController.java`: resolver 사용하도록 변경
   - `AuthorizedProductArgumentResolver.java`: 구현(리팩토링)
   - `ResolverConfig.java`: 일부 변경
   - `ProductOptionController.java`: 파라미터 변경(사용하지 않는 파라미터 언더스코어 처리)
- RoleUtil.java: 삭제

- 리팩토링에 따른 테스트 코드 변경
- 일부 테스트 코드 `@DisplayName` 추가

TODO: 테스트 코드에 `ResolverConfig`로 인해 mocking된 `JwtUtil`이 있는 것 같으나, 아직 확인하지 못함. 필요없으면 지우기

Gyuchool commented on src/main/java/gift/controller/user/ProductOptionController.java:
```diff
@@ -0,0 +1,66 @@
+import org.springframework.web.bind.annotation.RestController;
+
+@RestController
+@RequestMapping("/api/products/{productId}/options")
```
productId를 받는데 사용하지를 않네요?

kimjhyun0627 replied:
서비스 리팩토링하며 `option`의 `productId`가 현재 접근하는 `productId`와 일치하는지 확인하는 로직을 추가했습니다 !

Gyuchool commented on src/main/java/gift/controller/user/ProductOptionController.java:
```diff
@@ -0,0 +1,66 @@
+    @GetMapping
+    public ResponseEntity<List<OptionResponse>> getOptions(
+            @CurrentRole Role _role,
+            @AuthorizedProduct Product product
```
AuthorizedProduct 리졸버의 역할은 무엇인가요?🤔
서비스마다 결국 productRepository.findById()로 검증하고 있어서 리졸버가 꼭 필요한지 싶네요!

kimjhyun0627 replied:
리팩토링 진행하겠습니다 !

Gyuchool commented on src/main/java/gift/service/product/option/ProductOptionServiceImpl.java:
```diff
@@ -0,0 +1,68 @@
+            throw new OptionAlreadyExistException();
+        }
+
+        ProductOption option = ProductOption.of(name, quantity);
```
```suggestion
        ProductOption option = ProductOption.of(name, quantity, product);
```
이렇게 생성하지않고 따로 메서드를 뺀 이유가 있으실까요?

kimjhyun0627 replied:
객체 생성과 테이블 관계 설정은 다른 영역이라고 생각해서,  조금 더 명시적으로 두기 위해 메서드를 분리했던것으로 기억합니다 !

kimjhyun0627 replied:
테스트 코드를 위해 기존 메서드도 남겨두고 오버로딩 진행했습니다!

Gyuchool replied:
```
public void assignTo(Product product) {
        this.product = product;
    }
```
해당 메서드는 이제 테스트에서만 사용하게 되었군요.! 
테스트코드에서만 사용하는 코드를 프로덕션 코드가 가지고 있는 것에 대해서 어떻게 생각하실까요?🙄

Gyuchool commented on src/main/java/gift/entity/product/option/ProductOption.java:
```diff
@@ -0,0 +1,81 @@
+        return quantity;
+    }
+
+    public OptionResponse toResponse() {
```
도메인에서 dto를 반환하면 어떤 점이 문제가 될까요?

kimjhyun0627 replied:
여러 개의 DTO를 사용하는 경우 문제가 발생할 것 같습니다..!
도메인이 DTO을 알 필요도 없군요...!
리팩토링 진행하겠습니다 !

Gyuchool commented on src/main/java/gift/service/product/option/ProductOptionServiceImpl.java:
```diff
@@ -0,0 +1,68 @@
+    @Override
+    @Transactional(readOnly = true)
+    public List<OptionResponse> getOptions(Long productId) {
+        productRepository.findById(new ProductId(productId))
```
findBy의 용도가 무엇일까요? 없으면 어떻게 되나요?

kimjhyun0627 replied:
기본적으로 인가 관련 로직으로 구상하였는데, 인가 관련 로직을 리졸버로 이양하는 과정에서 `findById`를 따로 리팩토링하지 않고 이중 인가처럼 구성된 것 같습니다.(사실 이중 인가도 아닌 상태였긴 하네요...사실상 테이블 관계 설정을 위한 DB 조회로 봐도 무방할 것 같습니다)
DM으로 말씀해주셨 부분 반영해 리졸버 삭제하고 `role`을 인자로 받아 인가 로직을 서비스 레이어 내부로 완전히 이동시킴으로써 `findById`를 조금 더 효율적으로 사용할 수 있도록 변경했습니다!

Gyuchool commented on src/main/resources/db/init/schema.sql:
```diff
@@ -41,3 +41,17 @@ CREATE TABLE WISH
+
+    CONSTRAINT fk_option_product
+        FOREIGN KEY (PRODUCT_ID)
+            REFERENCES PRODUCT (ID) ON DELETE CASCADE
```
시간이 남았으니 미션과 조금 별개로 말씀드리면 CASCADE는 안쓰시는 것을 추천드려요!
아시다시피 product삭제할때 productoption까지 물리 삭제하다보니 편리하지만 너무 강력해서 조심해야하죠..!

Gyuchool commented on src/main/java/gift/service/product/option/ProductOptionServiceImpl.java:
```diff
@@ -0,0 +1,68 @@
+
+    @Override
+    public OptionResponse addOption(Long productId, String name, int quantity) {
+        Product product = productRepository.findById(new ProductId(productId))
+                .orElseThrow(() -> new ProductNotFoundException(productId));
+
+        if (
+                optionRepository.existsByProduct_IdAndName_Name(
+                        new ProductId(productId),
+                        name
+                )
+        ) {
+            throw new OptionAlreadyExistException();
+        }
```
더 효율적으로 DB IO에 접근하려면 어떻게 개선할 수 있을까요?
어렵게 생각하지 말아주세요 ㅎㅎ☺️

kimjhyun0627 replied:
`save()`와 `DataIntegrityViolationException`을 사용해 DB IO 횟수를 1회 줄인 것 같습니다..!
학습 도중 `getReferenceById()`라는 것도 알게 되었는데, 잘 모르고 쓰는 것보다는 조금 더 공부하고 써보고 싶어 일단은 다음으로 남겨두었습니다 !

Gyuchool replied:
리뷰에서 의도했던 것은 Option이 이미 존재하는지를 먼저 검증한다면, 매번 product를 조회하는 db io를 줄일 수 있을 것 같아서 남겼습니다:)

kimjhyun0627 replied:
오...! 감사합니다 !!
그런데 해당 상품에 대한 옵션이 중복된다는 것을 먼저 알고 처리를 해 버린다면, DB IO 횟수는 감소할 수 있지만, 접근 권한이 없는 상품에 대해 "상품을 찾을 수 없다" 가 아닌, "해당 옵션이 이미 존재한다"라는 정보가 반환되게 되어, 사용자가 "어? 이 상품이 있긴 하다고?" 식의 정보 누출이 발생하지 않나요..?
현재 의도적으로 "상품이 실제로 존재하지 않을 때" 와 "숨긴 상품을 권한 없이 CRUD하려고 할 때" 모두 NOT FOUND를 반환하고 있어 여쭤봅니다 !

Gyuchool replied:
예외를 누구에게 전달하느냐에 따라 처리 방식이 달라질 수 있다고 생각해요.
클라이언트(개발자나 내부 시스템)에게 전달하는 경우에는, 디버깅과 원인 파악을 위해 예외 메시지를 명확하게 내려주는 게 더 좋다고 봅니다.
반면, 최종 사용자(고객)에게 직접 노출되는 경우에는 말씀하신 것처럼 불필요한 내부 정보가 공개되지 않도록 예외 메시지를 제한하거나 추상화해야 한다고 생각해요.

그래서 만약 지금처럼 고객에게 예외 메시지가 직접 전달되는 구조라면, DM으로 말씀드렸듯이 커스텀 예외로 감싸서 적절한 메시지로 변환해 내려주는 방식이 필요하다고 생각해요

추가로 예외 핸들링과 DB I/O 중 우선순위를 따지자면, 저는 DB 접근 횟수를 줄이는 게 훨씬 더 중요한 최적화 포인트라고 생각해요.😊

Gyuchool reviewed (changes requested):
안녕하세요 진현님!
이번에도 빠르게 구현해주셨군요💯

테스트가 깨져서 수정후 재요청 부탁드릴게요🙏

<img width="294" height="154" alt="스크린샷 2025-07-20 오후 10 28 37" src="https://github.com/user-attachments/assets/69e4c741-e30d-49a4-95a1-9c64a52c03eb" />

몇 가지 코멘트를 남겨두었으니 반영 및 의견 주시고 다시 요청주세요!🙏
궁금한 점은 디스코드 DM이나 PR 커멘트로 언제든지 여쭤봐주세요!(새벽도 가능!)
끝까지 화이팅입니다💪

kimjhyun0627 committed:
refactor: Step 3 2차 리뷰 반영

- ProductOptionController.java: option의 productId가 현재 접근하는 productId와 일치하는지 확인하는 로직을 추가
- ProductOptionController.java: 인가 resolver 삭제
- ProductOption.java: ProductOption.of 오버로딩
- OptionResponse.java: ProductOption.toResponse 이양
- ProductOptionServiceImpl.java: 인가 관련 비즈니스 로직 이양
- ProductOptionServiceImpl.addOption: save()사용으로 DB 접근 횟수 감소
- schema.sql: CASCADE -> RESTRICT 변경
- 관련 테스트 코드 수정
- ResolverConfig 변경 -> 컨트롤러 MVC 테스트 시 불필요한 bean 삭제

kimjhyun0627 commented:
### 리뷰와 어제 DM으로 여쭤본 내용 바탕으로 리팩토링 진행했습니다 !

- `ProductOptionController.java`: `option`의 `productId`가 현재 접근하는 `productId`와 일치하는지 확인하는 로직을 추가

- `ProductOptionController.java`: 인가 `resolver` 삭제

- `ProductOption.java`: `ProductOption.of()` 오버로딩

- `OptionResponse.java`: `ProductOption.toResponse()` 이양 → `OptionResponse.of()`
- `ProductOptionServiceImpl.java`: 인가 관련 비즈니스 로직 이양
- `ProductOptionServiceImpl.addOption()`: `save()`사용으로 DB 접근 횟수 감소
- `schema.sql`: `CASCADE` -> `RESTRICT` 변경
- 관련 테스트 코드 수정
   - 로직 변경에 따른 수정 
   - `ResolverConfig.java` 변경 -> 컨트롤러 MVC 테스트 시 불필요한 `@MockitoBean` 삭제

Gyuchool commented on src/main/java/gift/service/product/option/ProductOptionServiceImpl.java:
```diff
@@ -0,0 +1,76 @@
+            ProductOption savedOption = optionRepository.save(option);
+            return OptionResponse.of(savedOption);
+        } catch (DataIntegrityViolationException ex) {
+            throw new OptionAlreadyExistException();
```
catch하고 예외를 던지는 의도가 무엇일까요? 결국 예외가 발생하는 건 똑같은게 아닌가 싶어서요!

참고하시면 좋을 예외처리 관한 글 공유드립니다:) https://jojoldu.tistory.com/734

kimjhyun0627 replied:
아직 걸어주신 링크를 읽지 않은 상태로 답변해보자면, 모호한 "데이터베이스 무결성 위반" 보다는 커스텀 예외로 어떤 부분에서 오류가 발생했는지를 알기 위해서 저렇게 구성했습니다 !

Gyuchool commented on src/main/java/gift/service/product/option/ProductOptionServiceImpl.java:
```diff
@@ -0,0 +1,76 @@
+        option.decreaseAmount(amount);
+    }
+
+    private Product authorizeProduct(Long productId, Role role) throws ProductNotFoundException {
```
throw 예외를 상위로 넘기신 이유가 있으실까요?

kimjhyun0627 replied:
오우...실수입니다 !!

Gyuchool commented on src/test/java/gift/controller/admin/AdminAuthControllerTest.java:
```diff
@@ -31,17 +32,17 @@
 @ExtendWith(SpringExtension.class)
 @WebMvcTest(AdminAuthController.class)
 @AutoConfigureMockMvc(addFilters = false)
+@Import(ResolverConfig.class)
```
Import는 무슨 용도이기에 적용하셨을까요? 제거하면 어떻게 되나요?

kimjhyun0627 replied:
리졸버 빈 등록을 위해서 붙인 어노테이션입니다! 없으면 빈 미등록 오류가 나서 500이 뜨더라구요...;)

Gyuchool reviewed (approved):
안녕하세요 진현님!
마지막 피드백도 빠르게 반영해주셨군요💯
한 주간 짧게나마 코드로 의견 주고받아서 즐거웠습니다😊 

몇 가지 코멘트를 남겨두었으니 다음 단계 진행하시기 전에 확인 및 의견 주시면 좋을 것 같습니다.🙏
궁금한 점은 디스코드 DM이나 PR 커멘트로 언제든지 여쭤봐주세요!(새벽도 가능!)
다음 미션도 지금처럼 화이팅입니다💪

kimjhyun0627 commented:
말씀해주신 부분 잘 수정해서 다음 과제에 반영하겠습니다.!
한 주 조금 넘는 시간동안 감사했습니다 !!
